presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java

```diff
@@ -0,0 +1,107 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+final class LifespanSerializedPageTracker
+        implements PagesReleasedListener
+{
+    private final OutputBufferMemoryManager memoryManager;
+    @Nullable
+    private final PagesReleasedListener childListener;
+    private final ConcurrentMap<Lifespan, AtomicLong> outstandingPageCountPerLifespan = new ConcurrentHashMap<>();
+    private final Set<Lifespan> noMorePagesForLifespan = ConcurrentHashMap.newKeySet();
+    private volatile Consumer<Lifespan> lifespanCompletionCallback;
+
+    public LifespanSerializedPageTracker(OutputBufferMemoryManager memoryManager)
+    {
+        this(memoryManager, Optional.empty());
+    }
+
+    public LifespanSerializedPageTracker(OutputBufferMemoryManager memoryManager, Optional<PagesReleasedListener> childListener)
+    {
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.childListener = requireNonNull(childListener, "childListener is null").orElse(null);
+    }
+
+    public boolean isLifespanCompletionCallbackRegistered()
+    {
+        return lifespanCompletionCallback != null;
+    }
+
+    public void registerLifespanCompletionCallback(Consumer<Lifespan> callback)
+    {
+        checkState(lifespanCompletionCallback == null, "lifespanCompletionCallback is already set");
+        this.lifespanCompletionCallback = requireNonNull(callback, "callback is null");
+    }
+
+    public void incrementLifespanPageCount(Lifespan lifespan, int pagesAdded)
+    {
+        AtomicLong counter = outstandingPageCountPerLifespan.get(lifespan);
+        if (counter == null) {
+            counter = outstandingPageCountPerLifespan.computeIfAbsent(lifespan, ignore -> new AtomicLong());
+        }
+        counter.addAndGet(pagesAdded);
+    }
+
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        requireNonNull(lifespan, "lifespan is null");
+        noMorePagesForLifespan.add(lifespan);
+    }
+
+    public boolean isNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        return noMorePagesForLifespan.contains(lifespan);
+    }
+
+    public boolean isFinishedForLifespan(Lifespan lifespan)
+    {
+        if (!noMorePagesForLifespan.contains(lifespan)) {
+            return false;
+        }
+
+        AtomicLong outstandingPageCount = outstandingPageCountPerLifespan.get(lifespan);
+        return outstandingPageCount == null || outstandingPageCount.get() == 0;
+    }
+
+    @Override
+    public void onPagesReleased(Lifespan lifespan, int releasedPageCount, long releasedSizeInBytes)
+    {
+        long outstandingPageCount = outstandingPageCountPerLifespan.get(lifespan).addAndGet(-releasedPageCount);
```

Comment:
nit: we can use `decrementAndGet(releasedPageCount)`

Reply:
`decrementAndGet()` takes no arguments as it implies a value of `-1`